### ISSUE TITLE
Zebra node establishes multiple inbound or outbound connections to the same peer

### ISSUE THREAD

**Issue by dimxy** [ZcashFoundation/zebra#6936]
### What happened?

I set up a small testnet with two nodes, zcash and zebra.
When I ask a zcash node to add a connection to a zebra node with an rpc `addnode 'ip:port' onetry` I expect that the zebra node creates only a single inbound connection to this peer. 

Instead, this happened:
After that inbound conn is set up, the zebra node establishes a second outbound connection to the same zcash peer node, by learning, I believe, its address from the version message.
I think, this is a sort of a problem as the second connection wastes a slot in the limited connection list.

When komodod or zcashd node gets an inbound connection they won't create an outbound connection to the same peer, I believe, thankfully to this code:
https://github.com/zcash/zcash/blob/fa3827656df8f97fb4700b22022cc581fbda61fa/src/net.cpp#L1663 (checking that there is no peer with this ip address already).
Maybe this check was not implemented deliberately in Zebra but I have not found any explanations for that so I'm creating this issue just for sure.

### What were you doing when the issue happened?

`zcash-cli addnode 'ip:port' onetry` to connect a zcash node to a zebra node

### Zebra Version

Zebra:1.0.0-rc.9 (build from the main branch on 12 June 2023)

### Which operating systems does the issue happen on?

- [ ] Linux
- [ ] macOS
- [ ] Windows
- [ ] Other OS

### OS details

not os dependent

### Additional information

Sample getpeerinfo output from a zcash node from a small testnet, with both inbound and outbound conns to a zebra node (7999 port is used for listening) :
```
./zcash-cli -testnet getpeerinfo
[
  {
    "id": 1,
    "addr": "XX.XX.87.135:7999",
    "addrlocal": "XX.XX.238.59:40006",
    "services": "0000000000000001",
    ...
    "version": 170100,
    "subver": "/Zebra:1.0.0-rc.9/",
     "inbound": false,
     ...
  },
  {
    "id": 2,
    "addr": "XX.XX.87.135:48982",
    "addrlocal": "XX.XX.238.59:7999",
    "services": "0000000000000001",
    ...
    "version": 170100,
    "subver": "/Zebra:1.0.0-rc.9/",
    "inbound": true,
     ...
  }
]
```

**Comment by mpguerra** [comment]
Thank you for taking the time to fill out a bug report! We'll triage it and decide how to proceed.

**Comment by arya2** [comment]
> After that inbound conn is set up, the zebra node establishes a second outbound connection to the same zcash peer node, by learning, I believe, its address from the version message.

It looks like it's getting the remote peer's address from the TCP request.

It's supposed to check that `!self.has_connection_recently_responded() && !self.was_connection_recently_attempted()` [here](https://github.com/ZcashFoundation/zebra/blob/main/zebra-network/src/meta_addr.rs#L572) before making an outbound connection. It updates the address book after negotiating the version, so I think it should set the `last_response` field of new addresses to the current time to disqualify new inbound connections as candidates for the outbound connector.

It's also supposed to drop an existing connection for that address if there is one when inserting a new connection into the peer set. I'm not sure if the outbound connector can reliably connect to the same address. We may want to consider adding a UUID to the `Version` message to store peer connections by their unique IDs instead.

**Comment by teor2345** [comment]
I'm not yet convinced this is a bug, and I don't think it's an urgent fix. We could wait until it's scheduled in a sprint to start work on it.

Multiple peers can run on the same machine, or behind the same NAT. This is common for IPv4 peers in countries or networks which have used up their IPv4 address allocations.

> > After that inbound conn is set up, the zebra node establishes a second outbound connection to the same zcash peer node, by learning, I believe, its address from the version message.
>
> It looks like it's getting the remote peer's address from the TCP request.

Addresses in version messages can be faked, so Zebra uses the address and port from the OS TCP network stack. For inbound connections, this is usually a temporary port which is randomly chosen by the remote peer's OS.

> It's supposed to check that `!self.has_connection_recently_responded() && !self.was_connection_recently_attempted()` [here](https://github.com/ZcashFoundation/zebra/blob/main/zebra-network/src/meta_addr.rs?rgh-link-date=2023-06-13T17%3A59%3A49Z#L572) before making an outbound connection. It updates the address book after negotiating the version, so I think it should set the `last_response` field of new addresses to the current time to disqualify new inbound connections as candidates for the outbound connector.

Does this fix introduce a security vulnerability where peers can interfere with other peers address book records, by using fake addresses or ports?

> It's also supposed to drop an existing connection for that address if there is one when inserting a new connection into the peer set.

The PeerSet should drop duplicate peers based on the address and port. We have logging for that since PR #6832, I've never seen the warning log for duplicate IP and port. Occasionally we have a duplicate IP with a different port.

> I'm not sure if the outbound connector can reliably connect to the same address.

Do you mean the same address as an address book entry, or something else?

> We may want to consider adding a UUID to the `Version` message to store peer connections by their unique IDs instead.

This is a network protocol change, I think it's out of scope for this bug fix. It also introduces a similar security bug to version addresses: peers can send fake UUIDs to change other peers address book entries.

**Comment by arya2** [comment]
> I'm not yet convinced this is a bug, and I don't think it's an urgent fix. We could wait until it's scheduled in a sprint to start work on it.

I agree (based off your comments in https://forum.zcashcommunity.com/t/ziggurat-3-0/43350/29), and I think we should wait until it's scheduled in a sprint.

> This is a network protocol change, I think it's out of scope for this bug fix. It also introduces a similar security bug to version addresses: peers can send fake UUIDs to change other peers address book entries.

I agree that it's out of scope for this bug fix, I think it could avoid the security bug with version addresses if it generates a UUID per edge because only the two connected nodes would have the connection ID.

**Comment by teor2345** [comment]
> > This is a network protocol change, I think it's out of scope for this bug fix. It also introduces a similar security bug to version addresses: peers can send fake UUIDs to change other peers address book entries.
> 
> I agree that it's out of scope for this bug fix, I think it could avoid the security bug with version addresses if it generates a UUID per edge because only the two connected nodes would have the connection ID.

Zcash connections are not encrypted, so any internet router could put the UUID from another peer in its own network message, and use it to perform an attack.

TLS already implements the kind of authentication you're looking for here, using server certificates.

**Comment by teor2345** [comment]
> > I'm not yet convinced this is a bug, and I don't think it's an urgent fix. We could wait until it's scheduled in a sprint to start work on it.
> 
> I agree (based off your comments in [forum.zcashcommunity.com/t/ziggurat-3-0/43350/29](https://forum.zcashcommunity.com/t/ziggurat-3-0/43350/29)), and I think we should wait until it's scheduled in a sprint.

I think there should be some limit on connections per IP address, to make some network attacks harder. But that limit doesn't have to be 1.

The address book doesn't track inbound connections, so we'd have to implement this limit in 2 places:
- the outbound connector should skip IP addresses that are already at the limit, before it does the handshake
- the peer set should drop new inbound or outbound connections to IP addresses that are already at the limit, after the handshake has been done

**Comment by teor2345** [comment]
The limit definitely shouldn't be 39 connections per IP:
> 2023-06-14T23:01:27.847018Z  INFO crawl: zebra_network::peer_set::set: duplicate IP addresses in peer set duplicate_connections=39 duplicated_peers=1 peers=39

https://github.com/ZcashFoundation/zebra/actions/runs/5269398827/jobs/9528968450?pr=6957#step:8:1441

**Comment by teor2345** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Engineering-Team-5fb24d9264a3e8000e666a9e/issues/ZcashFoundation/zebra/6936?planning-poker) with Zenhub @arya2 @dconnolly @oxarbitrage @upbqdn

**Comment by arya2** [comment]
> I think there should be some limit on connections per IP address, to make some network attacks harder.

Good idea, we could:
- Add a field to the peer set, `peers_by_ip: HashMap<IpAddr, Vec<PeerSocketAddr> | usize>`
- Update the field in [the `poll_discover()` method here](https://github.com/ZcashFoundation/zebra/blob/main/zebra-network/src/peer_set/set.rs#L483)
- Define a `MAX_CONNS_PER_IP` and immediately drop additional connections to a single IP

**Comment by teor2345** [comment]
> > I think there should be some limit on connections per IP address, to make some network attacks harder.
> 
> Good idea, we could:
> 
>     * Add a field to the peer set, `peers_by_ip: HashMap<IpAddr, Vec<PeerSocketAddr> | usize>`

This seems like a good long-term design. But adding an extra field means we need to keep it up to date every time we add or drop peers. Which needs a more detailed review and more tests.

But for a quick security fix, what about:
* Document that `MAX_CONNS_PER_IP = 1` for now
* Check `ready_services.keys()` and `unready_services.keys()` every time a peer is added in `poll_discover()`, and drop the new peer if its IP is in either list
* Update the tests that fail after this fix, splitting each one into two tests:
    * existing test data with duplicate IPs, updated asserts that make sure duplicate IPs are rejected
    * updated test data with unique IPs, existing asserts that make sure unique IPs are accepted

The peer set is limited to 200 peers by default, so the `keys()` iterations won't be expensive. (Network latency is slower by orders of magnitude.)

Then as follow-up tickets, we could:

Medium priority, important to prevent some kinds of DoS:
- Ignore repeated inbound connections from the same IP within the last N seconds, so we don't do a handshake
- Don't initiate handshakes to outbound IPs that were recently connected according to the AddressBook

Lower priority / less urgent:
- Change the limit to more than 1, using your design

**Comment by mpguerra** [comment]
This got closed when #6980 and #6983 were merged... should it have been?

**Comment by mpguerra** [comment]
Also @arya2  can you please add an estimate to this issue?

**Comment by teor2345** [comment]
> This got closed when #6980 and #6983 were merged... should it have been?

Those PRs fix the bug where Zebra makes multiple connections and holds them open for a long time, by dropping those connections in the peer set after the handshake.

#6981 and #6982 defend against specific kinds of denial of service attacks, by stopping duplicate connections before the handshake (and before they reach the peer set). But they're not urgent, we can schedule them some time in the next few sprints.

**Comment by teor2345** [comment]
@dimxy thanks for reporting this bug, it should be fixed in Zebra v1.0.1.

Please upgrade to the latest version:
https://github.com/ZcashFoundation/zebra/releases/tag/v1.0.1
https://hub.docker.com/r/zfnd/zebra/tags
